### DISCUSSION TITLE
Problems with Statamic Install

### DISCUSSION THREAD

**Discussion by mikeschinkel** [statamic/cms#10487] [Q&A] [answered]
Hi all,

I have been trying to get Statamic installed and working, but I keep running into problems I cannot get passed. 

The Quick Start suggested I come here and ask, so that's what I am doing; hoping that someone here can help me.

I am trying to install Statamic in a Docker container of my own so I can host in Portainer in my VMware ESXi server and then use lsyncd or Mutagen to sync the source files between my laptop and the VM.

I posted all the relevant configuration files I am using to this Gist: https://gist.github.com/mikeschinkel/cc5bfdff3bc3d35794f0551460965391

The first problem I have — though not a blocker — is that the assets are being served as **HTTP instead of HTTPS**.  From line 8 of the view source:

```
<link rel="preload" as="style" href="http://statamic.local/build/assets/site-9bd4acd3.css" />
<link rel="modulepreload" href="http://statamic.local/build/assets/site-4ed993c7.js" />
<link rel="stylesheet" href=ttp://statamic.local/build/assets/site-9bd4acd3.css" />
<script type="module" src="http://statamic.local/build/assets/site-4ed993c7.js"></script>
```
The second problem — which is a blocker — is I get this error this when I try to access the control panel:

```
Vite manifest not found at: /var/www/html/public/vendor/statamic/cp/build/manifest.json 
(View: /var/www/html/vendor/statamic/cms/resources/views/partials/head.blade.php)...
```

There is a `/var/www/html/public/build/manifest.json` on the server, but for some reason it wants to load from the directory `/var/www/html/public/vendor/statamic/cp/build/` instead.

There is also a message that says run `npm run dev` but the Quick Start says that all that is needed is to run `composer create-project` so I feel like I must have done something wrong.  

I did try to get `npm install && npm run dev` to work but it fails saying I have _"too many listeners"_ and I tried `npm install && npm run build` which did not fail to build the Docker container but resulted in same problems as above.

I then tried three (3) different starter kits — `stillat/the-agency`, `hettiger/stenciljs-starter-kit`, and `studio1902/peak` — and all of those introduced even more `npm` issues so I gave up on them. 

I just want to get Statamic up so I can see what it is like and be able to use xdebug to learn Laravel and Statamic. 

Note that I have lots of PHP and other backend coding experience (Golang, for example) but zero Laravel or Statamic experience and very little front-end dev experience before trying to get this working. 

I really don't want to give up on Statamic, but I am at a loss for how to get a working site at this point. So thank you in advance for any help on this.

-Mike

**Comment by sjclark** [comment]
> I just want to get Statamic up so I can see what it is like and be able to use xdebug to learn Laravel and Statamic.

😄 Definitely sounds like you're going for a fairly complex initial approach - if you're just testing it out then ideally you'd want to just fire up something local (such as Laravel Herd or Valet) rather than throwing all sorts of potential npm, environment, docker and nginx issues into the mix. The Quick Start references some more in-depth guides: https://statamic.dev/installing - really the quick-start is much more geared for local development.

In saying all that, have you checked your `APP_URL` is set correctly in your `.env`. And can you provide the output from running `npm install`? If you can't get npm working you're going to have bigger issues in any case.

**Reply by mikeschinkel** [reply]
I appreciate that advice, and if my circumstance were different I'd probably happily follow it.  

But I running locally is currently problematic because my machine is already running slowly and low on memory. I need to switch to a newer laptop, but I'm waiting for something to be able to do that.  

I am also taking the approach I am to be able to gain expertise in setting up Lavarel on Docker containers, as I am trying to kill two birds with one stone. I hope you don't mind that I am taking this approach. It seems to me that getting the correct PHP and assets is the part I am struggling with, and not the environment, Docker of nginx issues. I really didn't expect it to be so hard to get something working. I guess my past experience working with WordPress really spoiled me.

I did use https://statamic.dev/installing/ubuntu for setting up my container, I was just referencing the Quick Start as it said all you need to do is run `composer create-project`. That is also effectively all that https://statamic.dev/installing/ubuntu says to do, other than the parts about setting up Linux, Docker, permissions, etc. which I am a lot more comfortable with doing.

Setting `APP_URL` does not seem to have any affect, positive or negative. I have been planning to add that to my script but since it wasn't fixing anything when I would set it from the terminal I just had not done that yet.

Here is the log for my most recent attempt at `npm install && npm run build` which AFAICT ran as expected(?).  

https://gist.github.com/mikeschinkel/b94843319831830ed410aab572aa7c5c

It created these files, which seems fine, except Statamic tries to load them from `public/vendor/statamic/cp/build/` instead of from `public/build/`, and for all my googling and ChatGPTing I can't find anything that explains the discrepancy:

```
public/build/manifest.json              0.28 kB │ gzip: 0.15 kB
public/build/assets/site-r5AqOYD2.css  18.08 kB │ gzip: 3.27 kB
public/build/assets/site-l0sNRNKZ.js    0.00 kB │ gzip: 0.02 kB
```
On the surface the problem seems easy:

1. Configure it to generate HTTPS URLs instead of HTTP URLs for assets, and 
2. Get it to look in `public/build/` instead of `public/vendor/statamic/cp/build/` for assets.

Is this not an easy fix that just takes knowledge of how to do it?

**Reply by edalzell** [reply]
ok, so if you want to do it that way, can you get Laravel going first, without Statamic?

**Reply by edalzell** [reply]
> Get it to look in public/build/ instead of public/vendor/statamic/cp/build/ for assets.

Nope, the former is for the front end, the latter is for the CP. Your issue is that, for some reason, those CP files aren't there (that's the `statamic:install` I had you try.

Can you show us the output of that command?

**Reply by sjclark** [reply]
> Is this not an easy fix that just takes knowledge of how to do it?

It's more that you've just thrown a whole bunch of variables that could be completely outside of Statamics control (as Jack referenced).

Statamic is largely just bolted onto Laravel, so it could be worth you trying to simply install Laravel instead, to try narrow down your problems.

I'm not too sure where you're setting `APP_URL` but just double-checking it should be set in the `.env` file in the root folder of your statamic install (along with many other variables Laravel and Statamic will need). It's often a culprit for issues with both SSL and assets not loading.

**Reply by mikeschinkel** [reply]
> ok, so if you want to do it that way, can you get Laravel going first, without Statamic?

I am sure I can try.  But since I started with Statamic I don't know how to do that yet.  I'll look into that.

**Reply by edalzell** [reply]
Do this: https://laravel.com/docs/11.x#creating-a-laravel-project.

**Reply by mikeschinkel** [reply]
@edalzell 

> > Get it to look in public/build/ instead of public/vendor/statamic/cp/build/ for assets.
> Nope, the former is for the front end, the latter is for the CP. Your issue is that, for some reason, those CP files aren't there 

Oh, okay. Thank you for explaining that.

Seems obvious, in hindsight.  OTOH, it is really easy to get overwhelmed when things a person is new to don't work. 

> That's the `statamic:install` I had you try.  Can you show us the output of that command?

Here's a gist showing that and the `composer create-project` before it:

- https://gist.github.com/mikeschinkel/22fdadb0d5ccc2a9dc72932e26f9e587

**Reply by mikeschinkel** [reply]
@sjclark 
> I'm not too sure where you're setting `APP_URL` but just double-checking it should be set in the `.env` file in the root folder of your statamic install (along with many other variables Laravel and Statamic will need). It's often a culprit for issues with both SSL and assets not loading.

To clarify, I am dropping into a terminal shell prompt and using `nano` to edit `/var/www/html/.env` where `/var/www/html/public` is my nginx web root.

**Reply by mikeschinkel** [reply]
@edalzell 
> Do this: https://laravel.com/docs/11.x#creating-a-laravel-project.

Okay, I was able to get that to work, no problem, given my experience trying to get Statamic working.  It is running as a Docker container in Portainer on a Debian VM hosted by VMware ESXI server in my homelab and serving from `https://laravel.local` with a TLS certificate.  

It only displays one page with external links, but then I assume that is what should be expected?

<img width="618" alt="image" src="https://github.com/user-attachments/assets/8e8ec438-7008-441e-a7ce-c0102ff6278f">

BTW, thank you for helping. I really appreciate it.

**Comment by jackmcdade** [comment]
I highly recommend just using Herd instead of Docker if you’re just trying to try Statamic out. all of your problems would disappear and we’d be able to help if they didn’t.  Docker is a black hole of infinite configurability and we have very little experience with it here. I hope you’re okay trying one more new thing in the mix!

**Reply by mikeschinkel** [reply]
@jackmcdade — Thank you for the quick reply.

First I have heard of Herd — no pun intended — but I just went and checked it out. 

I may try out Herd, but I immediately see at least three (3) problems with it for my use-case. 

1.) Long story, but I am still on a 2015 MacBook Pro and it is showing its age (constantly running hot and slowing down) and 16 GB is just not enough memory to run PhpStorm and a VM.  I plan to switch soon, but that's not something I can do today.

2.) Herd is paid for debugging, and the reason I have time to work on this is because I am currently not on a paying project, and for that same reason I don't want to spend any money I don't have to.

3.) Ultimately I am looking for something to host in my Homelab VMware ESXi server in a container on a VM, so using Herd would mean I can't achieve that goal, at least not initially.

4.) The problem seems to be unrelated to Docker and all that other stuff. 

As I said in the other reply, it's really just these:

1. Configure it to generate HTTPS URLs instead of HTTP URLs for assets, and
2. Get it to look in public/build/ instead of public/vendor/statamic/cp/build/ for assets.

**Reply by edalzell** [reply]
> Herd is paid 

You don't need paid Herd, I don't have it and don't plan to pay. I use it for all my sites.

**Reply by edalzell** [reply]
> The problem seems to be unrelated to Docker and all that other stuff.

I'm not sure that's true, that's why we're trying to remove that as a variable.

**Reply by mikeschinkel** [reply]
> You don't need paid Herd, I don't have it and don't plan to pay. I use it for all my sites.

Does it support xdebug for the free version? 
Maybe I misunderstood what they meant by _"debugging capabilities"_ on [their pricing page](https://herd.laravel.com/#plans)?

> I'm not sure that's true, that's why we're trying to remove that as a variable.

Understood.

**Reply by martyf** [reply]
> > You don't need paid Herd, I don't have it and don't plan to pay. I use it for all my sites.
> 
> Does it support xdebug for the free version? Maybe I misunderstood what they meant by _"debugging capabilities"_ on [their pricing page](https://herd.laravel.com/#plans)?

You don't need Pro to use Xdebug:
https://herd.laravel.com/docs/1/advanced-usage/xdebug
https://laravel-news.com/xdebug-laravel-herd

Herd Pro makes it easier and all baked in to the UI (plus other features).

**Reply by mikeschinkel** [reply]
@martyf — Gotcha.  Thank you for clarifying my misconception.

I may try Herd to see if I can run on my machine without too much of a performance hit, but ultimately I want to have it running in a container on my ESXi server, so I still want to figure out why I can't seem to get it working there.

**Reply by mikeschinkel** [reply]
> I highly recommend just using Herd instead of Docker if you’re just trying to try Statamic out. 

So I decided to at least give Herd a try.  I followed [the instructions for installing Statamic with Herd](https://statamic.dev/installing/laravel-herd) and immediately hit a brick wall string to install Statamic CLI which seems like a problem with dependencies in `composer.json`, no?

```shell
$ composer global require statamic/cli
Xdebug: [Step Debug] Could not connect to debugging client. Tried: 127.0.0.1:9003 (fallback through xdebug.client_host/xdebug.client_port).
Changed current directory to /Users/mikeschinkel/.composer
./composer.json has been updated
Running composer update statamic/cli
Loading composer repositories with package information
Updating dependencies
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - symfony/console v4.3.3 requires php ^7.1.3 -> your php version (8.3.9) does not satisfy that requirement.
    - psy/psysh v0.9.9 requires symfony/console ~2.3.10|^2.4.2|~3.0|~4.0 -> satisfiable by symfony/console[v4.3.3].
    - psy/psysh is locked to version v0.9.9 and an update of this package was not requested.

You can also try re-running composer require with an explicit version constraint, e.g. "composer require statamic/cli:*" to figure out if any version is installable, or "composer require statamic/cli:^2.1" if you know which you need.

Installation failed, reverting ./composer.json and ./composer.lock to their original content.
```
![image](https://github.com/user-attachments/assets/924b466c-b505-4c63-996b-0d354658dc2b)

I am running PHP `8.3.9` and Composer `2.7.7` — both installed with `brew` — if that makes a difference.

**Reply by edalzell** [reply]
Yup you've got some older packages in there. If you're not using them, you can get rid of everything from that composer.json file (in `~/.composer) and start again.

Composer is a big part of Laravel dev so it's a good skill to get.

FYI, this is really nothing to do w/ Statamic, this is a composer issue. You'd have the same issue trying to install any global composer package.

**Reply by mikeschinkel** [reply]
Nevermind. I just realized there was an old, old version of `psy/psych` installed globally on my machine years ago, and after removing it I was able to install Statamic CLI.   But, have to leave for a doctor's appointment so will have to look at this later.

**Reply by mikeschinkel** [reply]
@edalzell — Thanks, our messages crossed. 🙂

**Reply by mikeschinkel** [reply]
@edalzell — 

> _Composer is a big part of Laravel dev so it's a good skill to get._

BTW, most of my experience with PHP was backend-dev for WordPress and writing plugins.  I developed a love-hate relationship with Composer because it was always causing dependency issues when I had to use it that I struggled to resolve in part because I did not use it often enough.

Back to today — too long of a story for now, but — my interest in Statamic is for a large Laravel-based CMS to work with so I could write some tools for PHP and Composer.  But, as I implied, that is a story for another day.

**Comment by mikeschinkel** [comment]
_OKAY_, after much pain and bullheaded effort, I was able to get it all working and figure out what caused the problems for future visitors and maybe even for updated Statamic documentation.

## Serving HTTPS URLs
First question was **_"How to force Statamic to serve HTTPS URLS?"_**

That one was easily answered and had nothing to do with using Docker vs. Herd.  I needed to add a call to `\Illuminate\Support\Facades\URL::forceScheme('https')`  in the `boot()` method in `./app/Providers/AppServiceProvider.php`. Doing that forces Statamic to always generated HTTPS URLs which I needed since I was running with an HTTPS cert.

Here is what that file looks like after I updated it:

```php
<?php  // <root>/app/Providers/AppServiceProvider.php

namespace App\Providers;

use Illuminate\Support\Facades\URL;
use Illuminate\Support\ServiceProvider;
use Statamic\Statamic;

class AppServiceProvider extends ServiceProvider
{
    /**
     * Register any application services.
     */
    public function register(): void
    {
        //
    }

    /**
     * Bootstrap any application services.
     */
    public function boot(): void
    {
        URL::forceScheme('https');
        // Statamic::vite('app', [
        //     'resources/js/cp.js',
        //     'resources/css/cp.css',
        // ]);
    }
}
```

**Reply by edalzell** [reply]
> That one was easily answered and had nothing to do with using Docker vs. Herd.

It actually does, because it's not needed in Herd. It happens when the proxy (in this case Docker) is connecting to app via port 80. When that happens then Laravels URL generation "guesses" the protocol and since it's port 80, it uses http. It can happen when you use the wrong SSL settings in Cloudflare or other proxies.

**Reply by edalzell** [reply]
Perhaps the Docker install docs could be improved though, thanks for figuring this out. You can PR a change into the docs if you have time.

**Reply by mikeschinkel** [reply]
@edalzell — From a technical standpoint, you are absolutely correct.  

And depending on what you envision Statamic to be, it may be an acceptable position to take, or it may be a damaging position for the goals of Statamic.

1. Does the team see Statamic merely as an add-on to Laravel, one where they expect users to first be completely immersed in all things Laravel before users even consider using Statamic?  If yes, then your point is well-taken.

2. Or does the team see Statamic as a CMS that _happens_ to be based on Laravel, and thus a plus for those who like Laravel but is generally an option for anyone who needs a CMS? Is Statamic intended to have its own life and its own user-base, even among those who are not significantly involved with Laravel? This frankly is the positioning that makes Statamic much more valuable as software vendor. 

If #2 is the vision for Statamic then taking the position that something _"is a Laravel thing"_ when that thing is a common requirement for users of Statamic — and what production websites don't need httpS these days? — then passing the buck on the Laravel is actively harmful for Statamic's adoption beyond hardcore Laravel users.

I spent 10 years working with WordPress — and 12 years before that as a CEO where I concerned myself with questions like _"Who is our target customer and how do we best serve them?"_ One thing I learned was that those most successful take full ownership of their user's experience and rarely if ever tell a user that their problem is because of some other vendor's product.  There are **many** reasons for WordPress' massive success in gaining market share, but one of the biggest reasons they succeeded is they bent over backwards to ensure nobody ever fails to get WordPress installed if there is anything possible they could have done to ensure the user did not run into problems. I never heard anyone in the WordPress community telling a user _"Well, that's a PHP thing"_ unless of course that user was a hardcore PHP developer already.

Again, market share may not be what the Statamic team are after. OTOH, since there is a paid Pro version, I rather doubt that Statamic does not care about growing its user base.  And one of the best things a software team can do to grow its user base is to make sure every user's initial experience has as little pain as possible. You will rarely find someone who spends as much time as I did to get around a problem. Hell, I normally would have just given up, but I am between clients so I decided to see if I could solve this and see if I could contribute what I learned back to Statamic.

Minimally I think Statamic should document how to force HTTPS and publish it in such a way that google will index it so people who google to learn how to force HTTS will quickly get their question answered.

Frankly I also think this should be a setting in the `.env` file and that when Statamic loads it should look for that setting and if found call `\Illuminate\Support\Facades\URL::forceScheme('https')` for the user. It is not like needing to serve only HTTPS URLs is an uncommon need.

But this is just one man's opinion. Take from it what you will.

**Reply by edalzell** [reply]
> It is not like needing to serve only HTTPS URLs is an uncommon need.

It's less common than you think because it's mostly dev-ops and every org is different.

But like I said, doc PRs are very VERY welcome.

**Reply by mikeschinkel** [reply]
@edalzell — Yeah, but how are dev-ops going to know how to do it, especially if they are not knowledgeable about Laravel?

BTW, take a look at [my LinkedIn](https://www.linkedin.com/in/mikeschinkel) and you'll find _"CI/CD DevOps"_ in my title. Clearly as a dev-op I did not know.

And yes, I was considering submitting at least one PR if it were welcomed, but let me first document the solution to the harder problem on this issue plus publish my code to GitHub, where I can point to a working `Dockerfile` and Docker compose before I can even think of cloning the docs and considering what I might add.

**Reply by edalzell** [reply]
I hear ya, it’s a tough issue because you would have had to do the same thing on a plain Laravel app, if you had assets, and we definitely don’t want to duplicate all of the info required to run a Laravel app.

**Reply by mikeschinkel** [reply]
@edalzell — I hear ya. and empathize. 

Maybe a moderate approach is to document those things that you find your users have issues with, and ignore those you don't hear about from your users?

BTW, in this case if you made an `.env` option for HTTPS generation then you wouldn't have to duplicate anything; you'd be documenting your own feature.

**Reply by edalzell** [reply]
Statamic doesn’t control the appserviceprovider, so I can’t really have anything in there. And you don’t really want to do that in a package because it’s hidden and you don’t know how folks have things set up. 

That’s why I suggested improving the docs via a PR.

**Reply by sjclark** [reply]
I think a big chunk of this is that realistically people handle http > https a thousand different ways. Can redirect in apache, nginx, caddy etc. Can forward ports. Can do it via `APP_URL`, can do by forcing it with the in-built Laravel mechanisms.

Every single example on the deployment page (Ploi, Forge, Docker Sail etc) handles it in a different way. Many people _do_ want http on the site itself because they'll have a proxy in front handling https. Theres just so many different bespoke configurations - so Statamic/Laravel largely take a naive approach.

Something to consider is that if it was set as an environment variable then while it might have worked for you it would have likely broken everyone elses setups where its largely auto-detecting. And could give the false impression that setting `https` magically means its now `https` where realistically theres a lot more moving parts (SSL certs etc) outside of Statamics control.

**Reply by mikeschinkel** [reply]
@sjclark — 

There are two sides to HTTP.  

1. Listening, serving, and/or redirecting, and 
2. Content generation. 

I am concerned about _content generation_, not about listening, serving, and/or redirecting _(my Traefik and Step-CA already handles that later, nicely.)_

> Every single example on the deployment page (Ploi, Forge, Docker Sail etc) handles it in a different way. 

Well, I definitely am not 100% familiar with how things are done in the Laravel world but in other web scenarios I am familiar with people rarely want to **_serve_** `http` URLs in their content because doing so requires 1.) stateful processing that otherwise does not need to be done, and 2.) a greater likelihood of serving content that is incorrect, i.e. replacing links that should be HTTP with HTTPS links. 

However, from what I have seen before, and recently, the Laravel world appears to be walled garden — much like the WordPress world that I am more familiar with — where things are handled the Laravel way in its own ecosystem in its own ways — Herd and Sail and Forge and Envoyer, etc. — and those ways often differ from how others not in the ecosystem do things. So maybe you are correct when it comes to Laravel users; they want to do it the Taylor Otwell way.  Of course, if you are interested in growing Statamic beyond existing Laravel users, expecting everyone to do it the Laravel way might not be a palatable approach beyond the Laravel faithful. 🤷‍♂️

> Something to consider is that if it was set as an environment variable then while it might have worked for you it would have likely broken everyone else's setups where its largely auto-detecting. And could give the false impression that setting https magically means its now https where realistically there's a lot more moving parts (SSL certs etc) outside of Statamics control.

I think most people setting up a web server know that generating HTTPS links does not make a site secure, especially considering the browser will tell them it is not secure the first time they visit such a page when there is no valid SSL certificate. So frankly I do not buy false impressions as a reasonable argument against making it easier to generate HTTPS in the content. #fwiw

That said, I have just been offering my perspective as I know it can be good for those on the inside to get an outsider's perspective from time to time because those on the inside can never know what it is like anymore to see their thing with a fresh set of eyes. But I am getting the sense that my input may not be appreciated, and if that is the case then just be explicit and I can move along.

**Reply by mikeschinkel** [reply]
@edalxell —
> Statamic doesn’t control the appserviceprovider, so I can’t really have anything in there. 

I assumed Laravel would allow developers to extend it, so I googled and found that Laravel documents how to [create your own providers](https://laravel.com/docs/11.x/providers).  I decided to do a proof-of-concept to see if in fact it would work and sure enough, it works exactly as I expected it to, side-by-side with an unmodified `AppServerProvider`:

```php
<?php

namespace App\Providers;

use Illuminate\Support\Facades\URL;
use Illuminate\Support\ServiceProvider;

class HttpsProvider extends ServiceProvider
{
    public function register(): void
    {
    }

    public function boot(): void
    {
        if (env('GENERATE_HTTPS', false)===true) {
            URL::forceScheme('https');
        }
    }
}
```
Is there some reason why Statamic could not add its own provider?

> And you don’t really want to do that in a package because it’s hidden and you don’t know how folks have things set up.

But couldn't the same argument be made against any setting in `.env`?   

Only speaking for myself, of course, but the first place I would look for settings — if I was trying to change something, or trying to understand why it behaves the way it does — would be in the settings file, and `.env` in the root is the first file I noticed when trying to find where settings were located.

> That’s why I suggested improving the docs via a PR.

As I said before, that's potentially in the cards too

**Reply by gymnae** [reply]
That snipped worked perfectly :)
I'm running statamic in a Proxmox Linux container running Caddy as web server, which itself sits behind another caddy, terminating TLS. Thus, only http traffic is exchanged between both caddy. I was unable to get caddy to rewrite all the responses to TLS protocol, and thus this hack is helpful. 
I hope this will survive updates.

**Comment by mikeschinkel** [comment, marked as answer]
## Missing Control Panel Assets including `manifest.json`

The second problem — the blocker — was much harder to track down and diagnose.

For the answer to the first problem — **Serving HTTPS URLs** — see [here](https://github.com/statamic/cms/discussions/10487#discussioncomment-10132956).
<hr>

<strong>TLDR</strong>; Running `composer create-project` in `docker build` disables plugins and thus fails to load [pixelfear/composer-dist-plugin](https://github.com/pixelfear/composer-dist-plugin) and thus fails to copy over the `dist` assets, thus causing the control panel to throw blocking errors.  

The solutions are to **_either_**: 

1. Set `ENV COMPOSER_ALLOW_SUPERUSER=1` before `RUN ...`, **_OR_**
2. Wrap either `RUN composer create-project ...` or `RUN statamic new ...` in `RUN script -q -c "<command goes here>"`.

Both of the above will stop Composer from disabling plugins during `docker build`.

<hr>

So after I install Statamic with Herd where everything worked I compared to when running in `docker build` where it did not. It was very hard to find the problem because it was masked by Composer disabling plugins. I had to track down numerous issues before I was finally able to pinpoint the reason for failure and then what would make it work.

I also noticed that the instruction for installing under Herd differ from installing on Ubuntu, and had they been the same I probably would have gotten around the problem sooner.  For Herd [the instructions are](https://statamic.dev/installing/laravel-herd):

```shell
composer global require statamic/cli
statamic new your-project-name
```

But for Ubuntu [the instructions are](https://statamic.dev/installing/ubuntu):

```shell
composer create-project --prefer-dist statamic/statamic example.com
```
After installing for Herd I went back to my Dockerfile and tried `statamic new your-project-name` but ran into a failed build with the error message _"TTY mode requires /dev/tty to be read/writable."_

![CleanShot 2024-07-24 at 01 01 56@2x](https://github.com/user-attachments/assets/1d90e2e7-dd6a-4b62-84fa-68b6dc6564df)

I was unable to track down what exactly caused this or how to resolve it, but I did figure out how to hack around it and get a successful result and that is by simulating a TTY with `script`, e.g.:

```php
script -q -c "statamic new --no-interaction statamic.local"
```
That got me a working Docker build that had all the required control panel assets.  And we can thank ChatGPT for suggesting the above, BTW.
<hr>

Still, even though I had a working `Dockerfile` I was frustrated that I did not yet understand what the underlying problem was. So I powered through it by debugging both with Composer and with Statamic CLI to discover what was failing, and why.

I was eventually able to track down the code that copies over the control panel assets that end up written to `./public/vendor/statamic/cp/build`.   Those are copied over from `./vendor/statamic/cms/resources/dist/build` when `php artisan statamic:install` is called inside Composer or Statamic CLI, and that ends us recursively calling `php artisan vendor:publish --tag statamic-cp --force true` which is what actually copies the files from `./vendor/...` to `./public/...`.

Of course the above assumes that the `pixelfear/composer-dist-plugin` composer plugin has downloaded [github.com/statamic/cms/releases/download/v5.17.0/dist.tar.gz](https://github.com/statamic/cms/releases/download/v5.17.0/dist.tar.gz) and untarred, either by Composer or indirectly by Statamic CLI.

_(BTW, how I wish I had known the above details two days ago, it would have saved me a day of debugging. I am posting here in detail in hopes to help the next person who runs into the same problems.)_

But the problem is that Composer — which generates quite a lot of output — just disables plugins and then carries on as it everything is _a-okay_, only outputting one line to indicate that it did not load `pixelfear/composer-dist-plugin`. I did not notice that error for a while because it did not say _"error"_ or _"warning"_ and I didn't realize that lack of `pixelfear/composer-dist-plugin` was the culprit until after I discovered that one-line message:

<img width="717" alt="CleanShot 2024-07-24 at 02 30 53@2x" src="https://github.com/user-attachments/assets/7611be21-83b6-46ff-8621-0eaf8167cb7c">

It wasn't until I found this message in the log that I was able to put 2 and 2 together:

```
Composer plugins have been disabled for safety in this non-interactive session.
Set COMPOSER_ALLOW_SUPERUSER=1 if you want to allow plugins to run as root/super user.
```
<img width="794" alt="CleanShot 2024-07-24 at 02 37 18@2x" src="https://github.com/user-attachments/assets/73f132b4-4ec3-4f09-af71-dc56051906a1">

So Composer is run without TTY in a non-interactive session during `docker build`, and when that happens Composer disables plugins for some reason.  That is unless you set the environment variable `COMPOSER_ALLOW_SUPERUSER` to `1` meaning you are running as `root` — which Composer [recommends against doing](https://getcomposer.org/root ) — but `root` it the standard user during a Docker build so in my opinion Composer's advice is moot for when running Composer inside of a `docker build`.

Or, in a `Dockerfile`:

```Dockerfile
ENV COMPOSER_ALLOW_SUPERUSER=1
```

Anyway, to repeat the advice given above in the TLDR;

The solutions are to **_either_**: 

1. Set `ENV COMPOSER_ALLOW_SUPERUSER=1` before `RUN ...`, **_OR_**
2. Wrap either `RUN composer create-project ...` or `RUN statamic new ...` in `RUN script -q -c "<command goes here>"`.

Both of the above will stop Composer from disabling plugins during `docker build` and they will thus allow `pixelfear/composer-dist-plugin` to download `github.com/statamic/cms/releases/download/v5.17.0/dist.tar.gz`, untar it to `./vendor/statamic/cms/resources/dist/build/manifest.json` and finally copy it to `./public/vendor/statamic/cp/build`.

<hr>

If you want to see the actual working files you can find them here:

- [Dockerfile.statamic](https://github.com/gearboxworks/portainer-stacks/blob/d762ce80398b343d35ced1834096d81a6a895aa2/Dockerfile.statamic) && [docker-compose.statamic.yaml](https://github.com/gearboxworks/portainer-stacks/blob/e53a892f5ef553496e65195025806c31f0c11c8e/docker-compose.statamic.yaml) && [Statamic support Files](https://github.com/gearboxworks/portainer-stacks/tree/main/files/statamic) _(copied over to container's root dir, e.g.`/`)_
- [Dockerfile.gbw-php8.3-fpm](https://github.com/gearboxworks/portainer-stacks/blob/d762ce80398b343d35ced1834096d81a6a895aa2/Dockerfile.gbw-php8.3-fpm) 
- [Dockerfile.traefik](https://github.com/gearboxworks/portainer-stacks/blob/899f456f9076db85d8362f2eae2367e01b38dcf2/Dockerfile.traefik) && [docker-composer.traefik.yaml](https://github.com/gearboxworks/portainer-stacks/blob/main/docker-compose.traefik.yaml) && [/etc/traefik/traefik.yaml](https://github.com/gearboxworks/portainer-stacks/blob/899f456f9076db85d8362f2eae2367e01b38dcf2/rootdir/etc/traefik/traefik.yaml)
- [docker-composer.step-ca.yaml](https://github.com/gearboxworks/portainer-stacks/blob/main/docker-compose.traefik.yaml) && [/etc/step/ca.json](https://github.com/gearboxworks/portainer-stacks/blob/899f456f9076db85d8362f2eae2367e01b38dcf2/rootdir/etc/step/ca.json)

Note that my Statamic install loads via HTTPS from `statamic.local` which is routed by [Traefik with Step-CA](https://mattedwards.org/2024/04/using-step-ca-certificates-with-traefik/) being the local certificate authority with `avahi` providing [mDNS](https://github.com/gearboxworks/portainer-stacks/blob/3243408693fb7b297ca6d6059db39e37c4580bc8/rootdir/usr/local/bin/mdns) for the local domains [using the Traefik API](https://github.com/gearboxworks/portainer-stacks/blob/3243408693fb7b297ca6d6059db39e37c4580bc8/rootdir/usr/local/bin/publish-traefik-mdns.sh) to keep track of [which local domain names](https://github.com/gearboxworks/portainer-stacks/blob/899f456f9076db85d8362f2eae2367e01b38dcf2/rootdir/etc/traefik/traefik-mdns.json) need to be served.

The [repo](https://github.com/gearboxworks/portainer-stacks/tree/d762ce80398b343d35ced1834096d81a6a895aa2) was designed to support a Portainer install on a VMware VM running Debian 11 in my homelab and as such is not configured to bootstrap itself completely. In the hopefully near future I will be replacing Portainer with K3s and scripting it 100% using Infrastructure as Code, but I will probably create a different named repo for that.

<hr>

All that said, hope these details help someone else in the future.

**Reply by jackmcdade** [reply]
Thank you so much for sticking with this process. I wish you didn't have so much trouble, but I'm working through improving the docs right now so hopefully it doesn't happen to someone else. 

You are mostly right in that I see Statamic as a CMS that _happens_ to be based on Laravel — but there is also a balance of not recreating the wheel with the enormous amount of things Laravel does provide for us. We used to simple defer to the Laravel docs in a lot of place, but over the years we've been duplicating/elaborating on their content in places to make it more clear and simple to use Statamic. 

I don't think it's possible to be all one way or other with it, but it is in _everyone's_ best interest to smooth out all the speed bumps possible. So again, thank you!

**Reply by jasonvarga** [reply]
I've added a page to the docs to help with this specific issue: https://statamic.dev/troubleshooting/missing-control-panel-assets

**Reply by mikeschinkel** [reply]
@jackmcdade & @jasonvarga — Thank you for your responses, and for updating the docs.

BTW, that new page is _EXCELLENT!_

Yes, I can see how much of an uphill climb it would be to duplicate all of Laravel's efforts — if for no other reason than how prolific Taylor Otwell is!  And I empathize.

Maybe the best way to think about it is to set up [mission and vision statements](https://www.atlassian.com/work-management/strategic-planning/mission-and-vision) — assuming you do not already have them — and then focusing your docs on achieve **_your_** vision, as opposed to duplicating Laravel's. 

More specifically, that means providing docs that target those who are building, deploying, maintaining, administering, and using a CMS vs. developing, deploying, and maintaining a PHP app. IOW, assuming your **mission** is _"To provide the best choice of CMS for \<your target market>"_ and your **vision** is for _"Every new CMS deployment by \<your target market> will be a Statamic install"_ then decide what actions to pursue and decisions to make based on how well they support your mission and vision.  Then your choices on what docs to provide/update will naturally be driven by and fall out of what fits your mission and vision.

If it supports your mission and vision but duplicates Laravel's efforts, then I suggest you still consider doing it. However, I suspect that if you make choices through that lens of mission and vision you will find that most often you won't need to be duplicating Laravel's efforts directly but instead uniquely catering to your desired audience.

-Mike 
P.S. _BTW, mission and vision should be way more specific than my examples, but I do not want to put the thumb on the scale for where you want to take Statamic so I left it generic. Also note that I am only making these suggestions in hopes my insight from my ~40 years of industry experience will be helpful to your future success. Take away from it what you will._

**Reply by mikeschinkel** [reply]
@jasonvarga — 

Hey, I noticed that your _"[_Recent Updates_](https://statamic.dev/)"_ section on your docs home page lists updated pages but not **new** pages like the one you linked above.  Were you aware of that and/or is that intended?  

Just thought I'd mention it if you were not aware.

**Comment by deathlylol** [comment]
I have the same problem with manifest.js 

Illuminate\Foundation\ViteManifestNotFoundException
Vite manifest not found at: /home/vagrant/code/bon_core/public/vendor/statamic/cp/build/manifest.json

I used this commands. npm install , npm run dev, npm build. But no one of them not helped for me. What do I need to do?

**Reply by duncanmcclean** [reply]
This guide might be helpful: https://statamic.dev/troubleshooting/missing-control-panel-assets
